### PR TITLE
Massively refactored to fix bugs and use double

### PR DESCRIPTION
**Pull request by Krellan**
Now uses floating-point double for all internal math
Converts to/from integer millidegrees only as needed during I/O
This solves many "off by a factor of 1000" bugs
This also makes offset/compensation/calibration more precise
Still accepting all 4 combinations: ["milli"]("degree" | "margin")

Now uses variants, as required by recent D-Bus schema
Checks for floating-point NAN and avoids accepting it
Still recognizing -1 in integer millidegrees as NAN equivalent

Now properly looks up appropriate service for each object path
Sensors are no longer limited to use only "Hwmon.external" service

Optional DEBUG constant enables printing useful line of logging
This line shows all margins considered, and the final worst margin chosen
Each zone has its own line, with zone ID number to tell them apart

Plugged file descriptor leak by closing all files successfully opened

JSON config file location may now be customized on command line

Cosmetic cleanup, such as removing trailing whitespace
Added a few more useful error messages
Added more temporary editor files to .gitignore list

Signed-off-by: Josh Lehan <krellan@google.com>

### CODE REVIEW

File: dbus/dbus.hpp [hunk 1]
```diff
@@ -26,29 +27,41 @@ class SDBusPlus
...
      * @param[in] objPath - object path.
-     * @param[in] value - value to be set.
+     * @param[in] value - value to be set, in degrees, regardless of unitMilli.
+     * @param[in] unitMilli - true to set millidegrees, false to set degrees.
      */
-    static auto
+    static void
         setValueProperty(sdbusplus::bus::bus& bus, const std::string& busName,
-            const std::string& objPath, const int64_t& value)
+            const std::string& objPath, const double& value, bool unitMilli)
```
Comment (alex310110) [inline dbus/dbus.hpp:35]: I've seen usage of double in other codes, but I'm still wondering if double is overkill, especially for processors without FPUs...

File: dbus/dbus.hpp [hunk 2]
```diff
@@ -26,29 +27,41 @@ class SDBusPlus
...
 
             methodCall.append(valueInterface);
             methodCall.append("Value");
-            methodCall.append(data);
+
+            // Using millidegrees on D-Bus seems to need int64_t not double
+            if (unitMilli)
+            {
+                // Convert degrees to millidegrees
+                auto intValue = static_cast<int64_t>(value * 1000.0);
```
Comment (alex310110) [inline dbus/dbus.hpp:49]: 1000 should be enough here instead of 1000.0 to avoid confusion?

File: dbus/dbus.hpp [hunk 3]
```diff
@@ -76,10 +92,36 @@ class SDBusPlus
...
+        {
+            // Receive as double
+            value = std::get<double>(propMap["Value"]);
         }
 
-        return std::get<double>(propMap["Value"]);
+        // Sensors in floating-point use nan to indicate error
+        if (std::isnan(value))
+        {
+            std::cerr << "Sensor Value reading not available: " << path << std::endl;
```
Comment (alex310110) [inline dbus/dbus.hpp]: Will it flood the log if the condition persists? Shall we print it only once?

File: main.cpp [hunk 4]
```diff
@@ -16,17 +16,17 @@ constexpr auto marginConfigPath =
...
     {
-        auto jsonData = parseValidateJson(marginConfigPath);
+        auto jsonData = parseValidateJson(configPath);
         sensorConfig = getSensorInfo(jsonData);
         skusConfig = getSkuInfo(jsonData);
     }
     catch (const std::exception& e)
     {
-        std::cerr << "Failed during building json file: " << e.what() << "\n";
+        std::cerr << "Failed during building json file: " << e.what() << std::endl;
```
Comment (alex310110) [inline main.cpp:29]: I'm still wondering when style on IO (https://github.com/openbmc/docs/blob/master/cpp-style-and-conventions.md#iostream) will be enforced... We may want to do it early in a separate commit if it's being accepted to OpenBMC.

File: util/util.cpp [hunk 5]
```diff
@@ -25,29 +29,40 @@ int getSkuNum()
...
 {
     auto bus = sdbusplus::bus::new_default();
     std::string service = getService(sensorDbusPath);
 
+    double value = std::numeric_limits<double>::quiet_NaN();
+
     if (service.empty())
     {
-        return -1;
+        std::cerr << "Sensor input path not mappable to service: " << sensorDbusPath << std::endl;
```
Comment (alex310110) [inline util/util.cpp]: Same questions on log flooding and the use of iostream.

File: util/util.cpp [hunk 6]
```diff
@@ -59,32 +74,56 @@ int getSpecTemp(struct conf::sensorConfig config)
...
     {
         sensorSpecFile.open(config.parametersPath, std::ios::in);
         if (sensorSpecFile)
         {
             sensorSpecFile >> specTemp;
+            sensorSpecFile.close();
         }
+    }
 
-        sensorSpecFile.close();
```
Comment (alex310110) [inline util/util.cpp:74]: I don't know what will happen if we close a file that is failed to open. If there's no negative effect, I kinda prefer to have only one close() call in the main body so that 1) it's cleaner; 2) prevent potential mistakes when someone is branching out the conditions in the future. If the std:fstream destructor can close the file by itself, we don't need this statement after all. Also a question on iostream, will C style file read be faster?

File: util/util.cpp [hunk 7]
```diff
@@ -59,32 +74,56 @@ int getSpecTemp(struct conf::sensorConfig config)
...
+    if (std::isnan(specTemp))
+    {
+        std::cerr << "Sensor MaxTemp reading not available: " << config.parametersPath << std::endl;
```
Comment (alex310110) [inline util/util.cpp]: Potential log flooding?

File: util/util.cpp [hunk 8]
```diff
@@ -59,32 +74,56 @@ int getSpecTemp(struct conf::sensorConfig config)
...
+    double targetOffset = static_cast<double>(targetOffsetInt);
+    targetOffset /= 1000.0;
+
+    double offsetvalue = 0.0;
     offsetvalue = setPoint / scalar;
-    if (targetTemp == -1)
+
+    // If targetTemp not specified, use maxTemp instead
+    double targetTemp;
+    if (targetTempInt == -1)
```
Comment (alex310110) [inline util/util.cpp:117]: Are we able to use more meaningful marker for things not specified instead of -1 at the early stage? Once this gets adopted by more companies, there will be more frictions to break the compatibility and clean the tech debt up.

File: util/util.cpp [hunk 9]
```diff
@@ -109,23 +148,32 @@ std::string getService(const std::string path)
     }
     catch (const sdbusplus::exception::SdBusError& ex)
     {
+        std::cerr << "Get dbus service fail. " << ex.what() << std::endl;
```
Comment (alex310110) [inline util/util.cpp]: Is getService() called in every loop? Looks expensive to me... Also potential log flooding?

File: util/util.cpp [hunk 10]
```diff
@@ -109,23 +148,32 @@ std::string getService(const std::string path)
...
         return "";
     }
 
     if (response.begin() == response.end())
     {
+        std::cerr << "Sensor service not found: " << path << std::endl;
```
Comment (alex310110) [inline util/util.cpp]: Ditto.

File: util/util.cpp [hunk 11]
```diff
@@ -109,23 +148,32 @@ std::string getService(const std::string path)
...
+void updateDbusMarginTemp(int zoneNum, double marginTemp, std::string targetpath)
 {
     auto bus = sdbusplus::bus::new_default();
-    std::string service = "xyz.openbmc_project.Hwmon.external";
+    std::string service = getService(targetpath);
 
-    dbus::SDBusPlus::setValueProperty(bus, service, targetpath, marginTemp);
+    if (service.empty())
+    {
+        std::cerr << "Sensor output path not mappable to service: " << targetpath << std::endl;
```
Comment (alex310110) [inline util/util.cpp]: Potential log flooding? Shall we refactor the code so that each sensor is wrapped up in a class, and if we get fundamental issue like this, we just remove this sensor from the polling loop?

File: util/util.cpp [hunk 12]
```diff
@@ -153,90 +201,171 @@ void updateMarginTempLoop(
     {
         for (int i = 0; i < numOfZones; i++)
```
Comment (alex310110) [inline util/util.cpp:202]: This loop code is too long to make variable i meaningful. Recommend to use variable names like zone_idx if possible.

File: util/util.cpp [hunk 13]
```diff
@@ -153,90 +201,171 @@ void updateMarginTempLoop(
...
+                if (std::isnan(sensorRealTemp))
                 {
-                    std::fstream sensorValueFile;
-                    sensorValueFile.open(sensorList[i][t->first].path, std::ios::in);
-                    if (sensorValueFile)
+                    // Sensor failure, unable to get reading
+                    if constexpr (DEBUG)
                     {
-                        sensorValueFile >> sensorRealTemp;
+                        std::cerr << " ?";
```
Comment (alex310110) [inline util/util.cpp:289]: Error message place holder?

File: util/util.cpp [hunk 14]
```diff
@@ -153,90 +201,171 @@ void updateMarginTempLoop(
...
+                    if ((std::isnan(calibMarginTemp)) ||
                         sensorRealTemp < calibMarginTemp)
                     {
                         calibMarginTemp = sensorRealTemp;
                     }
+
+                    // Sensor successful, already in margin format
+                    if constexpr (DEBUG)
+                    {
+                        std::cerr << " " << sensorRealTemp;
```
Comment (alex310110) [inline util/util.cpp:308]: Error message place holder?

File: util/util.cpp [hunk 15]
```diff
@@ -153,90 +201,171 @@ void updateMarginTempLoop(
...
-                    sensorCalibTemp += offsetVal;
-                    sensorCalibTemp *= sensorList[i][t->first].parametersScalar;
-
-                    if (calibMarginTemp == -1 ||
-                        sensorCalibTemp < calibMarginTemp)
+                    // Sensor failure, needed to know sensorSpecTemp to compute margin
+                    if constexpr (DEBUG)
                     {
-                        calibMarginTemp = sensorCalibTemp;
+                        std::cerr << " ?";
```
Comment (alex310110) [inline util/util.cpp:319]: Error message place holder?

File: util/util.cpp [hunk 16]
```diff
@@ -153,90 +201,171 @@ void updateMarginTempLoop(
...
+                if ((std::isnan(calibMarginTemp)) ||
+                    sensorCalibTemp < calibMarginTemp)
+                {
+                    calibMarginTemp = sensorCalibTemp;
+                }
+
+                // Sensor successful, converted absolute to margin
+                if constexpr (DEBUG)
+                {
+                    std::cerr << " " << sensorCalibTemp;
```
Comment (alex310110) [inline util/util.cpp:358]: Error message place holder?

File: util/util.cpp [hunk 17]
```diff
@@ -153,90 +201,171 @@ void updateMarginTempLoop(
...
+            {
+                calibMarginTemp = 0;
+            }
+
             updateDbusMarginTemp(i, calibMarginTemp, skuConfig[i].first.second);
+
+            // Finish sensor line, indicate computed worst margin
+            if constexpr (DEBUG)
+            {
+                std::cerr << " => " << calibMarginTemp << std::endl;
```
Comment (alex310110) [inline util/util.cpp:373]: Error message place holder?

**Review by alex310110 — CHANGES_REQUESTED** [review]
Thanks for putting this up!

File: dbus/dbus.hpp (hunk 1 above)
Comment (Krellan) [reply dbus/dbus.hpp:35]: Thanks, but the code already uses double in other places, so it's not a new addition here. This program already requires the kernel to do correct FPU emulation on ARM, but that's unchanged from how it was before. Passing in a double preserves accuracy, as it can then be converted "just in time" to integer at the last moment, after all other processing is done.

File: dbus/dbus.hpp (hunk 2 above)
Comment (Krellan) [reply dbus/dbus.hpp:49]: I've been burned before by integer constants triggering unwanted conversion to doing integer arithmetic, so when there's a constant in an expression, I want to make sure it has a decimal point, so the compiler keeps it as double. That way, a value of 0.5 will be scaled up to 500 appropriately here, instead of being prematurely rounded off then scaled wrong.

File: dbus/dbus.hpp (hunk 3 above)
Comment (Krellan) [reply dbus/dbus.hpp]: Good point. If a sensor is repeatedly failing, though, we would want to know about it. I can add some kind of throttling/flagging system in the future, though, to squelch a spammy sensor that is already known bad.

File: main.cpp (hunk 4 above)
Comment (Krellan) [reply main.cpp:29]: Agreed. This program also doesn't have a clang-format file yet, so the format is kind of random over time, as various people have worked on it. This was one of the few usages of "\n" so wanted to standardize it on std::endl which is what the rest of the program was using. Agreed that formatting cleanups should be in a separate commit, probably when it's ported into OpenBMC (if that happens). I wanted to keep this patch in one big patch, just to simplify it for early review before we have Gerrit.

File: util/util.cpp (hunk 5 above)
Comment (Krellan) [reply util/util.cpp]: Yep, makes sense here, same concern.

File: util/util.cpp (hunk 6 above)
Comment (Krellan) [reply util/util.cpp:74]: Unfortunately, GitHub doesn't let you see as much surrounding context as Gerrit does, when reviewing patches. The close() was moved to within the block that contains the successful open(). That way, we aren't trying to close a non-open file descriptor, and we're also remembering to close it before returning. There's at least one file descriptor leak bug that was fixed by doing these.

File: util/util.cpp (hunk 7 above)
Comment (Krellan) [reply util/util.cpp]: Same rationale as the sensor value reading above. It's important to see, but long-term you're right, should throttle/flag repeated errors from the same sensor, to make them less spammy.

File: util/util.cpp (hunk 8 above)
Comment (Krellan) [reply util/util.cpp:117]: The use of -1 to indicate "not available" or "doesn't apply here" in an integer field is already strongly existing convention, in this program. It's in README.md and example/example.json files. So, I'm adding nothing new here, merely following along with convention. By testing for this early, before any scaling or conversion to double is done, I'm also fixing bugs in which the -1 was wrongly getting through and being scaled up to -1000 millidegrees.

File: util/util.cpp (hunk 9 above)
Comment (Krellan) [reply util/util.cpp]: You're right. Good catch. I need to add a simple cache for the results of getService(), as it is extremely unlikely a sensor with the same name will change services at runtime.

File: util/util.cpp (hunk 11 above)
Comment (Krellan) [reply util/util.cpp]: Perhaps good ideas for the future. This patch was already getting big, though. Adding log throttling on a per-sensor basis would be the groundwork for adding the feature of temporary sensor disqualification, as you mentioned. Disqualified sensors would be polled much less often, so if they block or do other weirdness while they are continuing to malfunction, this won't be as painful.

File: util/util.cpp (hunk 12 above)
Comment (Krellan) [reply util/util.cpp:202]: Agreed. It isn't my doing, though. To change it from "i" would have a ripple effect that would make this patch even larger than it is. A general refactoring should still be done, though, to get nice cleanups like this. To do it now would bloat this patch, though, and obfuscate the genuine changes I'm trying to call attention to.

File: util/util.cpp (hunk 14 above)
Comment (Krellan) [reply util/util.cpp:308]: This, and all the similar comments, are part of building up a line of output. If you run the full loop until you get to the end, it will build up output like this: "Margin Zone 0: 90 86 90 85 58.875 78.875 58.5 60.75 55.312 57.437 57.062 60.312 55.437 56.562 => 55.312"
In this we can see all of the incoming margins from the temperature sensors, and the worst margin decision that was made. If an individual sensor malfunctions, it will show up as "?" instead of a number, so the total number of items in the output will always remain the same.

File: dbus/dbus.hpp (hunk 1 above)
Comment (alex310110) [reply dbus/dbus.hpp:35]: Yep, I just wonder if double is a bad choice from the beginning in all other components. :/

File: dbus/dbus.hpp (hunk 2 above)
Comment (alex310110) [reply dbus/dbus.hpp:49]: Oops, sry, I missed "const double& value" in the function signature. You are right about this.

File: util/util.cpp (hunk 14 above)
Comment (alex310110) [reply util/util.cpp:308]: I see. Can we construct this logging with a string (either sstream or std::string::operator+=, but I'm not sure which one has less performance overhead), so that this log will not be interrupted by some log message in some function call in the middle, and it's more readable?

**Comment by alex310110** [comment]
Updated my comments. Definitely some comments that I made earlier require another refactoring effort than this pull request. I hope we are getting some time some day to deal with them...

File: dbus/dbus.hpp (hunk 1 above)
Comment (Krellan) [reply dbus/dbus.hpp:35]: About the only disadvantage I would see is performance (since floating-point requires software emulation to work on ARM processor). It seems fine to me, because the main loop only runs once per second. CPU load is 4% to 7% on read-margin-temp and I suspect most of that is D-Bus overhead, not because of double. It would be worth benchmarking, though.

File: util/util.cpp (hunk 14 above)
Comment (Krellan) [reply util/util.cpp:308]: Yes, this sounds entirely reasonable. Makes sense, as in case somebody wants to add some debugging, it will interrupt the current line being built up. Should build up an internal string then log the entire line at once. Will add it as a future refinement.

File: util/util.cpp (hunk 6 above)
Comment (harveyquta) [reply util/util.cpp:74]: Hi Krellan,
Thanks to improvet this repo. It seems fix some bugs that I didn't find before.
About this situation, is it possible that use exception std::fstream::failure to catch error when file open failed?

File: util/util.cpp (hunk 9 above)
Comment (harveyquta) [reply util/util.cpp]: Maybe can print this log when DEBUG?

File: util/util.cpp (hunk 5 above)
Comment (harveyquta) [inline util/util.cpp]: Maybe can print this log when DEBUG?

File: util/util.cpp (hunk 9 above)
Comment (Krellan) [reply util/util.cpp]: Thanks for looking at this. This is an important error message, though, and should be printed if there's something wrong, lest the user wonder why their temperature sensor isn't showing up in the margin computation.

Comment (Krellan) [reply util/util.cpp]: I will take another look at this, though, and perhaps add some throttling to the log message. It could be annoying if printed more than once. It's important to still print it during normal execution, not DEBUG, though, because it represents a serious configuration error if getService() not found.

File: util/util.cpp (hunk 5 above)
Comment (Krellan) [reply util/util.cpp]: Thanks. I can see how this would be annoying if it repeats printing. It still needs to be printed during normal execution, not DEBUG, because it is an important error. I will add a throttling timer to it, similar to the getService() fix I will do.

File: util/util.cpp (hunk 6 above)
Comment (Krellan) [reply util/util.cpp:74]: You're welcome. That is something I will look at in the future. I don't think there's a need to use exceptions, the stream can be tested immediately after opening, using std::ios::operator bool(), to detect failure to open, which is already done here. There is something missing, though, an error message if the file could not be opened. This could be added, throttled like the others should be.

**Comment by harveyquta** [comment]
Hi Krellan,
I will pull request a commit first to improve some legibility. I think it doesn't conflict with your commit too much. You can see the changes in branch "dev_modified_save_data_method" or the new pull request. Thanks.

**Comment by Krellan** [comment]
OK, sounds good to me.

**Comment by harveyquta** [comment]
Hi Krellan, could you commit code first without those std::cerr log flooding? Thanks.

**Comment by harveyquta** [comment]
Hi Krellan, because we want to merge this pull request, could I comment those flooding log first and merge this PR?
Thanks a lot.

**Comment by Krellan** [comment]
Hi! Yes, that would be OK to do. Then, I can create a new pull request
after this has been merged, to properly fix the logging spam.

Josh

Josh Lehan | Software Engineer | krellan@google.com | +1 650-733-8941
